aten/src/ATen/native/cpu/FlashAttentionKernel.cpp

```diff
@@ -381,6 +446,28 @@ void cpu_flash_attention_backward(
   int64_t num_head = query.size(2);
   int64_t headSize = query.size(3);
 
+  bool has_attn_mask = attn_mask.has_value() && attn_mask.value().numel();
+  if (has_attn_mask) {
+    if (is_reduced_type) {
+      attn_mask.value() = attn_mask.value().to(at::kFloat);
+    }
+    int64_t attn_mask_size_0 = 1;
+    int64_t attn_mask_size_1 = 1;
+    if (attn_mask.value().dim() == 3
+        && attn_mask.value().size(0) == batchSize * num_head) {
+      attn_mask_size_0 = batchSize;
+      attn_mask_size_1 = num_head;
+    } else if (attn_mask.value().dim() == 4) {
+      if (attn_mask.value().size(0) == batchSize) {
+        attn_mask_size_0 = batchSize;
+      }
+      if (attn_mask.value().size(1) == num_head) {
+        attn_mask_size_1 = num_head;
+      }
+    }
+    attn_mask.value() = attn_mask.value().view({attn_mask_size_0, attn_mask_size_1, qSize, kvSize});
+  }
```

Comment:
Perhaps, you can factor out some util function to avoid duplicate code between forward and backward?

Reply:
Sure! A new function `reshape_attn_mask_to_4d` is created and the supported mask shapes are commented there.